### PR TITLE
RavenDB-10732: use smapsReader static instance for low mem

### PR DESCRIPTION
**Pull request by aviviadi**

### CODE REVIEW

File: src/Sparrow/LowMemory/LowMemoryNotification.cs [hunk 1]
```diff
@@ -325,7 +326,9 @@ public bool IsLowMemory(MemoryInfoResult memInfo, out long sharedCleanInBytes)
         {
             if (PlatformDetails.RunningOnLinux)
             {
-                var result = new SmapsReader().CalculateMemUsageFromSmaps<SmapsReaderNoAllocResults>();
+                if (_smapsReader == null) // allocate only in linux
```
Comment (ayende) [inline src/Sparrow/LowMemory/LowMemoryNotification.cs]: This is open to race conditions

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Sparrow/LowMemory/LowMemoryNotification.cs (hunk 1 above)
Comment (ayende) [reply src/Sparrow/LowMemory/LowMemoryNotification.cs]: Use a _local variable_, pass it a function argument.

File: src/Raven.Server/NotificationCenter/Handlers/ServerDashboardHandler.cs [hunk 2]
```diff
@@ -22,7 +24,8 @@ public class ServerDashboardHandler : ServerNotificationHandlerBase
 
                     try
                     {
-                        var machineResources = MachineResourcesNotificationSender.GetMachineResources();
+                        SmapsReader smapsReader = PlatformDetails.RunningOnLinux ? new SmapsReader(new[] {new byte[SmapsReader.BufferSize], new byte[SmapsReader.BufferSize]}) : null;
```
Comment (ayende) [inline src/Raven.Server/NotificationCenter/Handlers/ServerDashboardHandler.cs]: This is temporary, use an ArrayPool here.

File: src/Raven.Server/Dashboard/MachineResourcesNotificationSender.cs [hunk 3]
```diff
@@ -44,7 +45,8 @@ public class MachineResourcesNotificationSender : BackgroundWorkBase
                 if (_watchers.Count == 0)
                     return;
 
-                var machineResources = GetMachineResources();
+                SmapsReader smapsReader = PlatformDetails.RunningOnLinux ? new SmapsReader(new[] {new byte[SmapsReader.BufferSize], new byte[SmapsReader.BufferSize]}) : null;
```
Comment (ayende) [inline src/Raven.Server/Dashboard/MachineResourcesNotificationSender.cs]: Temporary instance, use a ArrayPool here.

File: src/Sparrow/LowMemory/LowMemoryNotification.cs [hunk 4]
```diff
@@ -315,20 +316,18 @@ internal int CheckMemoryStatus()
...
-            var isLowMemory = IsLowMemory(memInfo, out _);
+            var isLowMemory = IsLowMemory(memInfo, out _, smapsReader);
 
             // memInfo.AvailableMemory is updated in IsLowMemory for Linux (adding shared clean)
             memStats = (memInfo.AvailableMemory, memInfo.TotalPhysicalMemory, memInfo.CurrentCommitCharge);
             return isLowMemory;
         }
 
-        public bool IsLowMemory(MemoryInfoResult memInfo, out long sharedCleanInBytes)
+        public bool IsLowMemory(MemoryInfoResult memInfo, out long sharedCleanInBytes, SmapsReader smapsReader)
```
Comment (ayende) [inline src/Sparrow/LowMemory/LowMemoryNotification.cs]: Move the out param to the end

File: src/Raven.Server/Dashboard/MachineResourcesNotificationSender.cs [hunk 5]
```diff
@@ -45,7 +46,15 @@ public class MachineResourcesNotificationSender : BackgroundWorkBase
                 if (_watchers.Count == 0)
                     return;
 
-                SmapsReader smapsReader = PlatformDetails.RunningOnLinux ? new SmapsReader(new[] {new byte[SmapsReader.BufferSize], new byte[SmapsReader.BufferSize]}) : null;
+                SmapsReader smapsReader = null;
+                if (PlatformDetails.RunningOnLinux)
+                {
+                    var buffer1 = ArrayPool<byte>.Shared.Rent(SmapsReader.BufferSize);
+                    var buffer2 = ArrayPool<byte>.Shared.Rent(SmapsReader.BufferSize);
+                    smapsReader = new SmapsReader(new[] {buffer1, buffer2});
```
Comment (ayende) [inline src/Raven.Server/Dashboard/MachineResourcesNotificationSender.cs]: No, you just caused the buffer to be used by more than a single thread

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Dashboard/MachineResourcesNotificationSender.cs (hunk 5 above)
Comment (aviviadi) [reply src/Raven.Server/Dashboard/MachineResourcesNotificationSender.cs]: Should I use ThreadLocal<ArrayPool> member for that use?
```
ThreadLocal<ArrayPool<byte>> _arrayPool = ArrayPool<byte>.Create();
.
.
.
Task DoWork { bufferX = _arrayPool.Rent(..)
```

**Pull request merged**